grpc/disruption_listener_service.go

```diff
@@ -0,0 +1,126 @@
+type EndpointAlteration struct {
+	TargetEndpoint   string
+	ErrorToReturn    string
+	OverrideToReturn string
+}
+
+var (
+	errorMap = map[string]codes.Code{
+		"OK":                  codes.OK,
+		"CANCELED":            codes.Canceled,
+		"UNKNOWN":             codes.Unknown,
+		"INVALID_ARGUMENT":    codes.InvalidArgument,
+		"DEADLINE_EXCEEDED":   codes.DeadlineExceeded,
+		"NOT_FOUND":           codes.NotFound,
+		"ALREADY_EXISTS":      codes.AlreadyExists,
+		"PERMISSION_DENIED":   codes.PermissionDenied,
+		"RESOURCE_EXHAUSTED":  codes.ResourceExhausted,
+		"FAILED_PRECONDITION": codes.FailedPrecondition,
+		"ABORTED":             codes.Aborted,
+		"OUT_OF_RANGE":        codes.OutOfRange,
+		"UNIMPLEMENTED":       codes.Unimplemented,
+		"INTERNAL":            codes.Internal,
+		"UNAVAILABLE":         codes.Unavailable,
+		"DATALOSS":            codes.DataLoss,
+		"UNAUTHENTICATED":     codes.Unauthenticated,
+	}
+)
+
+// SendDisruption makes a gRPC call to DisruptionListener
+func (d *DisruptionListener) SendDisruption(ctx context.Context, ds *pb.DisruptionSpec) (*emptypb.Empty, error) {
+	if ds == nil {
+		log.Fatal("Cannot execute SendDisruption when DisruptionSpec is nil")
+	}
+
+	for _, alt := range ds.EndpointAlterations {
+		if alt.TargetEndpoint == "" {
+			log.Print("SendDisruption does not specify TargetEndpoint for all endpointAlterations")
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying TargetEndpoint for all endpointAlterations")
+		}
+		if alt.ErrorToReturn == "" && alt.OverrideToReturn == "" {
+			log.Print("For at least one endpointAlteration, SendDisruption specifies neither ErrorToReturn nor OverrideToReturn")
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying either ErrorToReturn or OverrideToReturn for all target endpoints")
+		}
+
+		alteration := EndpointAlteration{
+			TargetEndpoint:   alt.TargetEndpoint,
+			ErrorToReturn:    alt.ErrorToReturn,
+			OverrideToReturn: alt.OverrideToReturn,
+		}
+		d.Spec = append(d.Spec, alteration)
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+// DisruptionStatus lists all configured endpoint alterations for DisruptionListener
+func (d *DisruptionListener) DisruptionStatus(context.Context, *emptypb.Empty) (*pb.DisruptionSpec, error) {
+	ds := pb.DisruptionSpec{EndpointAlterations: []*pb.EndpointAlteration{}}
+
+	for _, alt := range d.Spec {
+		ds.EndpointAlterations = append(
+			ds.EndpointAlterations,
+			&pb.EndpointAlteration{
+				TargetEndpoint:   alt.TargetEndpoint,
+				ErrorToReturn:    alt.ErrorToReturn,
+				OverrideToReturn: alt.OverrideToReturn,
+			},
+		)
+	}
+	return &ds, nil
+}
+
+// CleanDisruption removes all configured endpoint alterations for DisruptionListener
+func (d *DisruptionListener) CleanDisruption(context.Context, *emptypb.Empty) (*emptypb.Empty, error) {
+	d.Spec = []EndpointAlteration{}
```

Comment:
I _think_ we should probably wrap this, as well everywhere else we set `d.Spec` with a mutex, to prevent multiple concurrent writers